modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogCompactionRunner.java

```diff
@@ -0,0 +1,343 @@
+     */
+    public CatalogCompactionRunner(
+            CatalogManagerImpl catalogManager,
+            MessagingService messagingService,
+            TopologyService topologyService,
+            LogicalTopologyService logicalTopologyService,
+            PlacementDriver placementDriver,
+            ClockService clockService,
+            Executor executor
+    ) {
+        this.messagingService = messagingService;
+        this.topologyService = topologyService;
+        this.logicalTopologyService = logicalTopologyService;
+        this.catalogManager = catalogManager;
+        this.clockService = clockService;
+        this.placementDriver = placementDriver;
+        this.executor = executor;
+    }
+
+    @Override
+    public CompletableFuture<Void> startAsync(ComponentContext componentContext) {
+        messagingService.addMessageHandler(CatalogMessageGroup.class, (message, sender, correlationId) -> {
+            assert message.groupType() == CatalogMessageGroup.GROUP_TYPE : message.groupType();
+            assert message.messageType() == CatalogMessageGroup.MINIMUM_REQUIRED_TIME_REQUEST : message.messageType();
+            assert correlationId != null;
+
+            long minimumRequiredTime = determineLocalMinimumRequiredTime();
+
+            CatalogMinimumRequiredTimeResponse response = MESSAGES_FACTORY.catalogMinimumRequiredTimeResponse()
+                    .timestamp(minimumRequiredTime)
+                    .build();
+
+            messagingService.respond(sender, response, correlationId);
+        });
+
+        return CompletableFutures.nullCompletedFuture();
+    }
+
+    @Override
+    public CompletableFuture<Void> stopAsync(ComponentContext componentContext) {
+        busyLock.block();
+
+        return CompletableFutures.nullCompletedFuture();
+    }
+
+    /** Updates the local view of the node with new compaction coordinator. */
+    public void updateCoordinator(ClusterNode newCoordinator) {
+        compactionCoordinatorNodeName = newCoordinator.name();
+
+        triggerCompaction(lowWatermark);
+    }
+
+    /** Returns local view of the node on who is currently compaction coordinator. For test purposes only.*/
+    @TestOnly
+    public @Nullable String coordinator() {
+        return compactionCoordinatorNodeName;
+    }
+
+    /** Called when the low watermark has been changed. */
+    public CompletableFuture<Boolean> onLowWatermarkChanged(HybridTimestamp newLowWatermark) {
+        lowWatermark = newLowWatermark;
+
+        triggerCompaction(newLowWatermark);
+
+        return CompletableFutures.falseCompletedFuture();
+    }
+
+    @TestOnly
+    CompletableFuture<Boolean> lastRunFuture() {
+        return lastRunFuture;
+    }
+
+    /** Starts the catalog compaction routine. */
+    CompletableFuture<Boolean> triggerCompaction(@Nullable HybridTimestamp lwm) {
+        if (lwm == null || !topologyService.localMember().name().equals(compactionCoordinatorNodeName)) {
```

Comment:
name of the node won't change during node lifetime. You don't need entire topologyService here, just pass nodeName in constructor

Reply:
Yes, I also originally wanted to do this, but the problem is that a non-started topologyService is passed to the constructor. I added a separate localMemberName field, which is initialized when the component starts.

Reply:
Got it,
fixed, thanks.